### PR TITLE
Added Windows WinRT/C++ support

### PR DESCRIPTION
**Pull request by jfkm69**
This adds supports for React Native Windows using WinRT/C++.

### CODE REVIEW

**Comment by namrog84** [comment]
This is awesome! I can't wait for it to get in!

File: package.json [hunk 1]
```diff
@@ -7,7 +7,8 @@
     "fast-base64-decode": "^1.0.0"
   },
   "peerDependencies": {
-    "react-native": ">=0.56"
+    "react-native": "^0.61.5",
+    "react-native-windows": "^0.61.15"
```
Comment (LinusU) [inline package.json:11]: Hmmm, do we really need to change the base react-native version here?

Also, now npm/yarn will complain about missing peer dependencies for people not using react-native-windows? Is it best practice to list it as a peer-dependency?

**Comment by LinusU** [comment]
At first look, this seems great. Will do a proper review soon, but it would be great if you could look at the peer dependency situation...

File: package.json (hunk 1 above)
Comment (acoates-ms) [reply package.json:11]: The base version shouldn't be changed just for windows.  Yes when using react-native-windows the required version will be >=0.61, but we shouldn't push that on all platforms.  Otherwise to use android/ios you'd suddenly need 0.61 which isn't true.  We should just document the requirement in the README.md or something.

The peer version for react-native-windows should be >=0.61.5, not ^ -- otherwise it won't work for 0.62 etc.

Unfortunately there isn't a npm concept of optionalPeerDependency or something.  The build of the windows module will require using react-native-windows, so it is a dependency when using the windows platform.

File: windows/RNGetRandomValues.vcxproj [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    <UseDebugLibraries>false</UseDebugLibraries>
+    <WholeProgramOptimization>true</WholeProgramOptimization>
+    <LinkIncremental>false</LinkIncremental>
+  </PropertyGroup>
+  <Import Project="$(VCTargetsPath)\Microsoft.Cpp.props" />
+  <ImportGroup Label="ExtensionSettings">
+  </ImportGroup>
+  <ImportGroup Label="Shared">
+    <Import Project="..\..\react-native-windows\Microsoft.ReactNative.Cxx\Microsoft.ReactNative.Cxx.vcxitems" Label="Shared" Condition="Exists('..\..\react-native-windows\Microsoft.ReactNative.Cxx\Microsoft.ReactNative.Cxx.vcxitems')" />
+    <Import Project="..\node_modules\react-native-windows\Microsoft.ReactNative.Cxx\Microsoft.ReactNative.Cxx.vcxitems" Label="Shared" Condition="Exists('..\node_modules\react-native-windows\Microsoft.ReactNative.Cxx\Microsoft.ReactNative.Cxx.vcxitems')" />
```
Comment (NickGerleman) [inline windows/RNGetRandomValues.vcxproj:76]: This directory won't always be correct, such as cases where the module is used in a monorepo. For RNW itself, we have some logic to go find this path (see ReactNativeWindowsDir in its build logic). 

Same below.

File: windows/RNGetRandomValues.vcxproj [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+      <Platform>Win32</Platform>
+    </ProjectConfiguration>
+    <ProjectConfiguration Include="Release|x64">
+      <Configuration>Release</Configuration>
+      <Platform>x64</Platform>
+    </ProjectConfiguration>
+  </ItemGroup>
+  <PropertyGroup Label="Configuration">
+    <ConfigurationType>DynamicLibrary</ConfigurationType>
+    <PlatformToolset>v140</PlatformToolset>
```
Comment (NickGerleman) [inline windows/RNGetRandomValues.vcxproj:56]: An FYI that this is likely to cause issues for folks using RNW 0.62 or up, since the build moved to v141.

Comment (NickGerleman) [reply windows/RNGetRandomValues.vcxproj:56]: nevermind, missed that this was redefined right after

File: windows/RNGetRandomValues.vcxproj [hunk 4]
```diff
@@ -0,0 +1,167 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project DefaultTargets="Build" ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$(SolutionDir)packages\Microsoft.Windows.CppWinRT.2.0.190730.2\build\native\Microsoft.Windows.CppWinRT.props" Condition="Exists('$(SolutionDir)packages\Microsoft.Windows.CppWinRT.2.0.190730.2\build\native\Microsoft.Windows.CppWinRT.props')" />
```
Comment (NickGerleman) [inline windows/RNGetRandomValues.vcxproj:3]: Having this point to the cppwinrt version in the SolutionDir might cause some issues, but there has been some active discussion on what we should be doing with these dependncies. @jonthysell might have a good idea here.

File: package.json (hunk 1 above)
Comment (jonthysell) [reply package.json:11]: Really, you don't even need to make react-native-windows a peerDependency if you don't want. At the end of the day, it's an app's dependency on rnw that will pull the package down.

File: windows/RNGetRandomValues.vcxproj (hunk 4 above)
Comment (jonthysell) [reply windows/RNGetRandomValues.vcxproj:3]: This is correct. This makes sure you can both build the lib in local development, and when an app tries to consume it, the project will still build because the package will be restored relative to the app solution

File: package.json (hunk 1 above)
Comment (NickGerleman) [reply package.json:11]: That's not quite always right though. If a peer dependency isn't declared, the package manager won't neccesarily put the 'react-native-windows' package in a place that's reachable by this package. Some package managers like PNPM explicitly create a special node_modules structure so that anything not declared as a dependency cannot be found.

So, if you were missing a peer dep, then were using PNPM, logic in the vcxproj to find react-native-windows wouldn't be able to find anything.

Similarly, you can have a dependency tree with two different versions of RNW, and if you don't have a peer dependency, a version of react-native-windows might not be made visible to this module, even if you're using yarn/npm.

Basically, it works 90% of the time, but will fail in certain cases if you don't include it.

File: windows/RNGetRandomValues.cpp [hunk 5]
```diff
@@ -0,0 +1,11 @@
+#include "pch.h"
+#include "RNGetRandomValues.h"
+
+namespace winrt::RNGetRandomValues::implementation
+{
+    hstring RNGetRandomValues::getRandomBase64(int byteLength)
```
Comment (asklar) [inline windows/RNGetRandomValues.cpp:6]: consider returning std::string (via conversion to utf8) instead of hstring to not leak WinRT-isms

File: .gitignore [hunk 6]
```diff
@@ -46,7 +46,15 @@ buck-out/
...
 #Windows
 .vs/
-Generated Files/
 packages/
+
+# Files generated by the VS build
+**/Generated Files/**
+
+# Windows Build output folders
+ARM/
```
Comment (namrog84) [inline .gitignore:55]: arm64 too

File: windows/RNGetRandomValues.sln [hunk 7]
```diff
@@ -0,0 +1,193 @@
+﻿
```
Comment (namrog84) [inline windows/RNGetRandomValues.sln:1]: This file isn't neccesary.

File: windows/RNGetRandomValues61/RNGetRandomValues61.vcxproj [hunk 8]
```diff
@@ -0,0 +1,166 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (namrog84) [inline windows/RNGetRandomValues61/RNGetRandomValues61.vcxproj:1]: Can you elaborate on the windows/RNGetRandomValues vs windows/RNGetRandomValues61 ?

Comment (jonthysell) [reply windows/RNGetRandomValues61/RNGetRandomValues61.vcxproj:1]: Unfortunately, 61-62 involved several tooling changes, and it would be a little complicated to have one project file that supported both. So even though the source files can be shared, having two projects, one for 61, and one for 62+ is a workaround pattern.

Going forward, especially once @jfkm69 switches to consuming the CppLib props/targets, will make the regular RNGetRandomValues.vcxproj better insulated against future changes to RNW.

File: windows/RNGetRandomValues/RNGetRandomValues.cpp [hunk 9]
```diff
@@ -0,0 +1,12 @@
+#include "pch.h"
+#include "RNGetRandomValues.h"
+
+namespace winrt::RNGetRandomValues::implementation
+{
+    REACT_METHOD(getRandomBase64)
```
Comment (namrog84) [inline windows/RNGetRandomValues/RNGetRandomValues.cpp:6]: Did you mean to move this back to the CPP? I think the REACT_METHOD is supposed to go in the header?

**Comment by namrog84** [comment]
Bump.  What all needs to happen for this to get in?

File: windows/RNGetRandomValues/RNGetRandomValues.vcxproj [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+    <UseDebugLibraries>false</UseDebugLibraries>
+    <WholeProgramOptimization>true</WholeProgramOptimization>
+    <LinkIncremental>false</LinkIncremental>
+  </PropertyGroup>
+  <Import Project="$(VCTargetsPath)\Microsoft.Cpp.props" />
+  <ImportGroup Label="ExtensionSettings">
+  </ImportGroup>
+  <ImportGroup Label="Shared">
+    <Import Project="$(ReactNativeWindowsDir)\Microsoft.ReactNative.Cxx\Microsoft.ReactNative.Cxx.vcxitems" Label="Shared" Condition="Exists('$(ReactNativeWindowsDir)\Microsoft.ReactNative.Cxx\Microsoft.ReactNative.Cxx.vcxitems')" />
+  </ImportGroup>
```
Comment (adamgorMSFT) [inline windows/RNGetRandomValues/RNGetRandomValues.vcxproj:79]: @jfkm69 line 77-79 can be deleted

File: windows/RNGetRandomValues/RNGetRandomValues.vcxproj [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+  </ItemGroup>
+  <ItemGroup>
+    <None Include="PropertySheet.props" />
+  </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="$(ReactNativeWindowsDir)\Microsoft.ReactNative\Microsoft.ReactNative.vcxproj">
+      <Project>{f7d32bd0-2749-483e-9a0d-1635ef7e3136}</Project>
+      <Private>false</Private>
+    </ProjectReference>
+  </ItemGroup>
```
Comment (adamgorMSFT) [inline windows/RNGetRandomValues/RNGetRandomValues.vcxproj:159]: line 154-159 here can be deleted.

RNW63 and newer I think only needs the .target and .props reference now and not the vcxitem and vcxproj items.

File: windows/RNGetRandomValues/RNGetRandomValues.vcxproj [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+    <ProjectConfiguration Include="Release|x64">
+      <Configuration>Release</Configuration>
+      <Platform>x64</Platform>
+    </ProjectConfiguration>
+  </ItemGroup>
+  <PropertyGroup Label="Configuration">
+    <ConfigurationType>DynamicLibrary</ConfigurationType>
+    <PlatformToolset>v140</PlatformToolset>
+    <PlatformToolset Condition="'$(VisualStudioVersion)' == '15.0'">v141</PlatformToolset>
+    <PlatformToolset Condition="'$(VisualStudioVersion)' == '16.0'">v142</PlatformToolset>
```
Comment (adamgorMSFT) [inline windows/RNGetRandomValues/RNGetRandomValues.vcxproj:61]: If you delete line 59-62, it should fallback to 'default platform toolset' which will support these versions AND VS2022.


Delete these lines.

    <PlatformToolset>v140</PlatformToolset>
    <PlatformToolset Condition="'$(VisualStudioVersion)' == '15.0'">v141</PlatformToolset>
    <PlatformToolset Condition="'$(VisualStudioVersion)' == '16.0'">v142</PlatformToolset>

File: windows/RNGetRandomValues/RNGetRandomValues.vcxproj [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+      <Configuration>Debug</Configuration>
+      <Platform>Win32</Platform>
+    </ProjectConfiguration>
+    <ProjectConfiguration Include="Debug|x64">
+      <Configuration>Debug</Configuration>
+      <Platform>x64</Platform>
+    </ProjectConfiguration>
+    <ProjectConfiguration Include="Release|ARM">
+      <Configuration>Release</Configuration>
+      <Platform>ARM</Platform>
```
Comment (adamgorMSFT) [inline windows/RNGetRandomValues/RNGetRandomValues.vcxproj:42]: ARM (32 bit) is no longer supported and the references to ARM should be removed.

ARM64 is still supported and should be kept.

**Comment by namrog84** [comment]
Hi @LinusU, I'd love to see this get in. 
Is there anything any of us can do to help to see this get in? 
Thanks!

**Review by 818Nawaf — APPROVED** [review]
